lib/Dialect/TTIR/Transforms/Passes.cpp

```diff
@@ -258,46 +258,126 @@ class TTIRGeneric : public impl::TTIRGenericBase<TTIRGeneric> {
+        op->getNumOperands(), rewriter.getAttr<OperandConstraintAttr>(
+                                  OperandConstraint::AnyDeviceTile)));
+    auto dispatch = rewriter.create<ttir::GenericOp>(
+        op.getLoc(), op->getResults().getTypes(), dps.getDpsInputs(),
+        dps.getDpsInits(), rewriter.getAttr<GridAttr>(), indexingMaps,
+        iteratorTypes, constraints);
 
-        blockArg.setType(layout.getMemref());
+    // Create a new basic block for the dispatch op and create block arguments
+    Block *block = rewriter.createBlock(&dispatch.getRegion());
+    SmallVector<Location> blockArgumentLocs(dispatch.getOperands().size(),
+                                            dispatch.getLoc());
+    block->addArguments(TypeRange(dispatch.getOperandTypes()),
+                        blockArgumentLocs);
 
-        bool isResult = dpsInterface.isDpsInit(&operand);
-        if (isResult) {
-          for (auto *user : blockArg.getUsers()) {
-            dyn_cast<KernelOp>(user).getResult(0).setType(layout.getMemref());
-          }
-        }
-      });
+    // Convert the original op into arith/math and into the dispatch block
+    OpBuilder blockBuilder = OpBuilder::atBlockEnd(block);
+    op.buildGenericRegion(blockBuilder, block);
+    rewriter.replaceOp(op, dispatch);
+    return success();
+  }
+};
+
+class TTIRGenericRegion
+    : public impl::TTIRGenericRegionBase<TTIRGenericRegion> {
+public:
+  using impl::TTIRGenericRegionBase<TTIRGenericRegion>::TTIRGenericRegionBase;
+  void runOnOperation() final {
+    RewritePatternSet patterns(&getContext());
+    patterns.add<TTIRGenericRegionRewriter>(&getContext());
+    FrozenRewritePatternSet patternSet(std::move(patterns));
+    if (failed(applyPatternsAndFoldGreedily(getOperation(), patternSet))) {
+      signalPassFailure();
+    }
+  }
+  void getDependentDialects(mlir::DialectRegistry &registry) const override {
+    registry.insert<mlir::tt::ttir::TTIRDialect>();
+    registry.insert<mlir::tt::TTDialect>();
+    registry.insert<mlir::arith::ArithDialect>();
+  }
+};
+
+struct TTIRGenericOperandsToMemrefRewriter
+    : public OpConversionPattern<GenericOp> {
+  using OpConversionPattern<GenericOp>::OpConversionPattern;
+
+  template <typename ValueRange>
+  void convertTypes(ValueRange valueRange,
+                    DenseMap<Type, Type> const &typeMap) const {
+    for (auto operand : valueRange) {
+      if (typeMap.count(operand.getType()) == 0) {
+        continue;
+      }
+      operand.setType(typeMap.at(operand.getType()));
     }
+  }
+
+  LogicalResult
+  matchAndRewrite(GenericOp generic, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    Block *entry = &generic.getRegion().front();
+    auto firstEntryArgType = entry->getArguments()[0].getType();
+    bool isConverted = static_cast<bool>(
+        mlir::cast<RankedTensorType>(firstEntryArgType).getEncoding());
+    if (isConverted) {
```

Comment:
You can embed getEncoding() call directly into if condition. isConverted isn't used afterwards.

Reply:
I deliberately wrote it this way so that the named bool `isConverted` tells you what the condition is looking for.  Imo just writing `if (mlir::cast<RankedTensorType>(firstEntryArgType).getEncoding()))` is a bit more implicit.  I can change it if you disagree.